In [0]:
from csv import reader
from pyspark.sql import Row 
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import warnings

import os
os.environ["PYSPARK_PYTHON"] = "python3"

In [0]:
import urllib.request
urllib.request.urlretrieve("https://data.seattle.gov/resource/tazs-3rd5.csv?$limit=1200000", "/tmp/my.csv")
dbutils.fs.mv("file:/tmp/my.csv", "dbfs:/seattle_crime_analysis/data/sea_crime.csv")
display(dbutils.fs.ls("dbfs:/seattle_crime_analysis/data/"))

path name size modificationTime dbfs:/seattle_crime_analysis/data/sea_crime.csv sea_crime.csv 295137527 1740171119000

In [0]:
data_path = "dbfs:/seattle_crime_analysis/data/sea_crime.csv"

In [0]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

sea_crime_df = spark.read.format("csv").option("header", "true").load(data_path)
display(sea_crime_df.limit(10))
sea_crime_df.createOrReplaceTempView("sea_crime")

report_number offense_id offense_start_datetime offense_end_datetime report_datetime group_a_b crime_against_category offense_parent_group offense offense_code precinct sector beat mcpp _100_block_address longitude latitude 2020-044452 12605598696 2020-02-03T08:00:00.000 2020-02-04T08:00:00.000 2020-02-05T10:06:28.000 A PROPERTY LARCENY-THEFT Theft of Motor Vehicle Parts or Accessories 23G N J J3 ROOSEVELT/RAVENNA 63XX BLOCK OF 5TH AVE NE -122.323399063 47.675117890 2020-044465 12605567653 2020-02-02T20:30:00.000 2020-02-02T21:30:00.000 2020-02-05T09:39:33.000 A PROPERTY ROBBERY Robbery 120 N U U3 ROOSEVELT/RAVENNA 26TH AVE NE / NE BLAKELEY ST -122.299552180 47.666384070 2020-044225 12605174036 2020-02-05T01:17:00.000 2020-02-05T02:21:00.000 2020-02-05T03:30:55.000 A PROPERTY DESTRUCTION/DAMAGE/VANDALISM OF PROPERTY Destruction/Damage/Vandalism of Property 290 W Q Q1 MAGNOLIA 22XX BLOCK OF W RAYE ST -122.384864805 47.642927340 2020-044076 12605081469 2020-02-05T00:51:21.000 null 2020-02-05T00:51:31.000 B SOCIETY DRIVING UNDER THE INFLUENCE Driving Under the Influence 90D N B B2 BALLARD SOUTH NW 46TH ST / 8TH AVE NW -122.366195342 47.662193080 2020-044142 12605077150 2020-02-04T22:59:00.000 null 2020-02-05T00:45:08.000 A PROPERTY LARCENY-THEFT Shoplifting 23C W M M1 DOWNTOWN COMMERCIAL 14XX BLOCK OF 2ND AVE -122.338449515 47.608788660 2020-044104 12605029468 2020-02-04T20:30:00.000 2020-02-04T20:45:00.000 2020-02-04T23:31:52.000 A PROPERTY DESTRUCTION/DAMAGE/VANDALISM OF PROPERTY Destruction/Damage/Vandalism of Property 290 N J J3 PHINNEY RIDGE 71XX BLOCK OF GREENWOOD AVE N -122.355309065 47.680526950 2020-043988 12604995759 2020-02-04T19:58:00.000 null 2020-02-04T22:46:19.000 A PROPERTY LARCENY-THEFT Shoplifting 23C SW F F2 ROXHILL/WESTWOOD/ARBOR HEIGHTS 26XX BLOCK OF SW BARTON ST -122.367311993 47.521027580 2020-043928 12604963106 2020-02-04T19:01:00.000 null 2020-02-04T21:59:07.000 B SOCIETY DRIVING UNDER THE INFLUENCE Driving Under the Influence 90D N L L2 NORTHGATE 15TH AVE NE / NE 95TH ST -122.312077769 47.697616170 2020-044065 12605008517 2020-02-04T21:00:00.000 2020-02-04T21:15:00.000 2020-02-04T21:47:32.000 A PROPERTY LARCENY-THEFT Theft From Motor Vehicle 23F E E E2 CAPITOL HILL 15XX BLOCK OF 12TH AVE -122.316845012 47.614683850 2020-044038 12604928711 2020-02-04T20:57:00.000 null 2020-02-04T21:20:35.000 A PROPERTY LARCENY-THEFT Shoplifting 23C E E E2 CAPITOL HILL 15XX BLOCK OF BROADWAY -122.320789514 47.614655490

### 1. Counting the number of crims for different category

In [0]:
count_result = sea_crime_df.groupBy('offense_parent_group').count().orderBy('count', ascending=False)
display(count_result)

offense_parent_group count LARCENY-THEFT 434101 ASSAULT OFFENSES 169334 BURGLARY/BREAKING&ENTERING 133721 DESTRUCTION/DAMAGE/VANDALISM OF PROPERTY 110111 MOTOR VEHICLE THEFT 82174 FRAUD OFFENSES 65078 TRESPASS OF REAL PROPERTY 35445 DRUG/NARCOTIC OFFENSES 29516 ROBBERY 27417 DRIVING UNDER THE INFLUENCE 20574 WEAPON LAW VIOLATIONS 11651 FAMILY OFFENSES, NONVIOLENT 10470 STOLEN PROPERTY OFFENSES 7822 SEX OFFENSES 7816 BAD CHECKS 6730 COUNTERFEITING/FORGERY 5764 PROSTITUTION OFFENSES 4147 LIQUOR LAW VIOLATIONS 2357 ARSON 2134 EMBEZZLEMENT 2091 KIDNAPPING/ABDUCTION 1707 EXTORTION/BLACKMAIL 1206 CURFEW/LOITERING/VAGRANCY VIOLATIONS 916 HOMICIDE OFFENSES 609 PORNOGRAPHY/OBSCENE MATERIAL 522 ANIMAL CRUELTY 231 SEX OFFENSES, CONSENSUAL 214 PEEPING TOM 205 HUMAN TRAFFICKING 84 GAMBLING OFFENSES 28 DRUNKENNESS 21 BRIBERY 18

Databricks visualization. Run in Databricks to view.

### 2. Counts the number of crimes for different district

In [0]:
loc_counts = sea_crime_df.groupBy("mcpp").count().orderBy('count', ascending=False)
display(loc_counts)


mcpp count DOWNTOWN COMMERCIAL 91534 CAPITOL HILL 72277 NORTHGATE 65864 QUEEN ANNE 60090 SLU/CASCADE 52159 UNIVERSITY 47121 ROOSEVELT/RAVENNA 42654 BALLARD SOUTH 38047 FIRST HILL 36502 LAKECITY 32386 CHINATOWN/INTERNATIONAL DISTRICT 31740 BELLTOWN 31200 CENTRAL AREA/SQUIRE PARK 30402 SANDPOINT 25696 GREENWOOD 25247 BITTERLAKE 23835 BALLARD NORTH 23691 FREMONT 21758 NORTH BEACON HILL 21443 WALLINGFORD 21208 SODO 20400 PIONEER SQUARE 19341 ROXHILL/WESTWOOD/ARBOR HEIGHTS 18679 MAGNOLIA 18354 ALASKA JUNCTION 17262 MOUNT BAKER 16980 BRIGHTON/DUNLAP 16675 MADRONA/LESCHI 15508 GEORGETOWN 15340 HIGHLAND PARK 13270 RAINIER BEACH 13027 MID BEACON HILL 12562 NORTH ADMIRAL 11510 JUDKINS PARK/NORTH BEACON HILL 10963 RAINIER VIEW 10908 MORGAN 10453 UNKNOWN 9964 HIGH POINT 9743 SOUTH PARK 9574 PHINNEY RIDGE 9569 NORTH DELRIDGE 9004 MILLER PARK 7931 MONTLAKE/PORTAGE BAY 7765 COLUMBIA CITY 7660 EASTLAKE - WEST 7593 NEW HOLLY 7326 CLAREMONT/RAINIER VISTA 6960 LAKEWOOD/SEWARD PARK 6305 ALKI 6067 HILLMAN CITY 5728 SOUTH BEACON HILL 5076 FAUNTLEROY SW 4862 SOUTH DELRIDGE 4808 MADISON PARK 3972 GENESEE 3696 EASTLAKE - EAST 1841 PIGEON POINT 1410 COMMERCIAL DUWAMISH 799 COMMERCIAL HARBOR ISLAND 445 null 23 NULL 4 1 CHINATOWN/INTERNATION DISTRICT 1 CASE/CASCAD 1

Databricks visualization. Run in Databricks to view.

### 3. Analysis the number of crime in each month of 2015, 2016, 2017, 2018. Then give insight about business impact for result.

In [0]:
from pyspark.sql.functions import col, to_timestamp, date_format, year, month, hour, to_date
sea_crime_df = sea_crime_df.withColumn(
    "formatted_report_datetime", 
    date_format(to_timestamp(col("report_datetime"), "yyyy-MM-dd'T'HH:mm:ss.SSS"), "yyyy-MM-dd HH:mm:ss")
) \
    .withColumn(
    "IncidentDate", 
    date_format(to_timestamp(col("report_datetime"), "yyyy-MM-dd'T'HH:mm:ss.SSS"), "MM/dd/yyyy")
)

sea_crime_df = sea_crime_df.withColumn("year", year(col('formatted_report_datetime'))) \
    .withColumn("month", month(col('formatted_report_datetime'))) \
    .withColumn("time", hour(col('formatted_report_datetime')))
display(sea_crime_df.limit(10))


report_number offense_id offense_start_datetime offense_end_datetime report_datetime group_a_b crime_against_category offense_parent_group offense offense_code precinct sector beat mcpp _100_block_address longitude latitude formatted_report_datetime IncidentDate year month time 2020-044452 12605598696 2020-02-03T08:00:00.000 2020-02-04T08:00:00.000 2020-02-05T10:06:28.000 A PROPERTY LARCENY-THEFT Theft of Motor Vehicle Parts or Accessories 23G N J J3 ROOSEVELT/RAVENNA 63XX BLOCK OF 5TH AVE NE -122.323399063 47.675117890 2020-02-05 10:06:28 02/05/2020 2020 2 10 2020-044465 12605567653 2020-02-02T20:30:00.000 2020-02-02T21:30:00.000 2020-02-05T09:39:33.000 A PROPERTY ROBBERY Robbery 120 N U U3 ROOSEVELT/RAVENNA 26TH AVE NE / NE BLAKELEY ST -122.299552180 47.666384070 2020-02-05 09:39:33 02/05/2020 2020 2 9 2020-044225 12605174036 2020-02-05T01:17:00.000 2020-02-05T02:21:00.000 2020-02-05T03:30:55.000 A PROPERTY DESTRUCTION/DAMAGE/VANDALISM OF PROPERTY Destruction/Damage/Vandalism of Property 290 W Q Q1 MAGNOLIA 22XX BLOCK OF W RAYE ST -122.384864805 47.642927340 2020-02-05 03:30:55 02/05/2020 2020 2 3 2020-044076 12605081469 2020-02-05T00:51:21.000 null 2020-02-05T00:51:31.000 B SOCIETY DRIVING UNDER THE INFLUENCE Driving Under the Influence 90D N B B2 BALLARD SOUTH NW 46TH ST / 8TH AVE NW -122.366195342 47.662193080 2020-02-05 00:51:31 02/05/2020 2020 2 0 2020-044142 12605077150 2020-02-04T22:59:00.000 null 2020-02-05T00:45:08.000 A PROPERTY LARCENY-THEFT Shoplifting 23C W M M1 DOWNTOWN COMMERCIAL 14XX BLOCK OF 2ND AVE -122.338449515 47.608788660 2020-02-05 00:45:08 02/05/2020 2020 2 0 2020-044104 12605029468 2020-02-04T20:30:00.000 2020-02-04T20:45:00.000 2020-02-04T23:31:52.000 A PROPERTY DESTRUCTION/DAMAGE/VANDALISM OF PROPERTY Destruction/Damage/Vandalism of Property 290 N J J3 PHINNEY RIDGE 71XX BLOCK OF GREENWOOD AVE N -122.355309065 47.680526950 2020-02-04 23:31:52 02/04/2020 2020 2 23 2020-043988 12604995759 2020-02-04T19:58:00.000 null 2020-02-04T22:46:19.000 A PROPERTY LARCENY-THEFT Shoplifting 23C SW F F2 ROXHILL/WESTWOOD/ARBOR HEIGHTS 26XX BLOCK OF SW BARTON ST -122.367311993 47.521027580 2020-02-04 22:46:19 02/04/2020 2020 2 22 2020-043928 12604963106 2020-02-04T19:01:00.000 null 2020-02-04T21:59:07.000 B SOCIETY DRIVING UNDER THE INFLUENCE Driving Under the Influence 90D N L L2 NORTHGATE 15TH AVE NE / NE 95TH ST -122.312077769 47.697616170 2020-02-04 21:59:07 02/04/2020 2020 2 21 2020-044065 12605008517 2020-02-04T21:00:00.000 2020-02-04T21:15:00.000 2020-02-04T21:47:32.000 A PROPERTY LARCENY-THEFT Theft From Motor Vehicle 23F E E E2 CAPITOL HILL 15XX BLOCK OF 12TH AVE -122.316845012 47.614683850 2020-02-04 21:47:32 02/04/2020 2020 2 21 2020-044038 12604928711 2020-02-04T20:57:00.000 null 2020-02-04T21:20:35.000 A PROPERTY LARCENY-THEFT Shoplifting 23C E E E2 CAPITOL HILL 15XX BLOCK OF BROADWAY -122.320789514 47.614655490 2020-02-04 21:20:35 02/04/2020 2020 2 21

In [0]:
sea_crime_df.createOrReplaceTempView("sea_crime_df")
sea_crime_1518 = spark.sql("""
    SELECT year, month, COUNT(*) AS count 
    FROM sea_crime_df 
    WHERE year >= 2015
    AND year <= 2018 
    GROUP BY year, month 
    ORDER BY count DESC""")
display(sea_crime_1518)

year month count 2018 5 6667 2018 7 6628 2017 8 6506 2018 8 6466 2017 10 6458 2017 7 6440 2018 10 6396 2018 6 6380 2018 12 6275 2018 9 6247 2018 1 6218 2017 3 6212 2018 11 6180 2016 5 6169 2017 5 6168 2017 1 6167 2018 3 6138 2016 7 6125 2016 1 6072 2017 12 6066 2018 4 6060 2017 6 6057 2017 9 6014 2016 10 5966 2016 12 5954 2016 11 5940 2017 4 5938 2015 10 5906 2016 8 5888 2015 8 5848 2017 11 5838 2015 12 5813 2015 11 5785 2016 2 5754 2015 6 5741 2016 6 5737 2015 9 5701 2015 5 5681 2015 7 5670 2016 9 5641 2018 2 5636 2015 4 5621 2015 3 5613 2016 3 5582 2016 4 5552 2015 1 5476 2017 2 5221 2015 2 5016

Databricks visualization. Run in Databricks to view.

### Crime Distribution
#### 1. Crime Distribution Over Time
High Crime Rate Months:

May: May 2018 (6,667 crimes) and May 2016 (6,169 crimes) had the highest crime rates.

July: July 2018 (6,628 crimes) and July 2016 (6,125 crimes) also saw high crime activity.

August: August 2017 (6,506 crimes) and August 2018 (6,466 crimes) were peak months for crime.

Low Crime Months:

February: February 2018 (5,636 crimes) and February 2015 (5,016 crimes) had the lowest crime rates.

January: January 2015 (5,476 crimes) and January 2016 (6,072 crimes) also saw relatively low crime activity.

#### 2. Annual Trends
2018: Overall crime rates were high, especially in May, July, August, and October.

2017: Crime rates were also high, particularly in August, October, and July.

2016: Crime rates were relatively stable, with peaks in May and July.

2015: Crime rates were generally lower, especially in February and January.

#### 3. Seasonal Trends
Spring and Summer (May - August): Crime rates were consistently higher, possibly due to warmer weather and increased outdoor activities.

Fall and Winter (October - December): Crime rates remained high, potentially linked to holidays like Halloween and Christmas.

Winter (January - February): Crime rates were lower, likely due to colder weather and reduced outdoor activities.

###Business Impact
#### 1. Resource Allocation
Peak Season Resource Deployment:

Increase police patrols during high-crime months (e.g., May, July, August) in crime hotspots like city centers and commercial areas.

Enhance community policing and public safety campaigns to raise awareness among residents.

Low Season Resource Optimization:

Use low-crime months (e.g., January, February) for officer training, equipment maintenance, and resource optimization.

#### 2. Preventive Measures
Seasonal Prevention Strategies:

During spring and summer, increase nighttime patrols and surveillance in high-risk areas like entertainment districts and parks.

During holidays, deploy additional police forces to prevent theft, robbery, and other crimes.

Data-Driven Decision Making:

Use historical crime data to predict future trends and implement proactive measures.

Identify crime hotspots through data analysis and implement targeted security measures.

## 4. Analyse time frequency in Seattle districts and provide advice for local police department.

In [0]:
sea_crime_df.createOrReplaceTempView("sea_crime_df")
sea_crime_hournday = spark.sql("""
    SELECT IncidentDate, time, COUNT(*) AS count 
    FROM sea_crime_df 
    WHERE IncidentDate IN ('12/15/2020', '12/15/2021', '12/15/2022')
    GROUP BY IncidentDate, time
    ORDER BY count DESC""")
display(sea_crime_hournday)

IncidentDate time count 12/15/2022 6 54 12/15/2021 13 34 12/15/2021 14 34 12/15/2021 16 28 12/15/2021 10 19 12/15/2021 15 19 12/15/2020 10 18 12/15/2021 12 17 12/15/2021 17 15 12/15/2021 18 15 12/15/2020 9 14 12/15/2020 13 13 12/15/2022 18 13 12/15/2022 17 13 12/15/2022 14 13 12/15/2021 22 13 12/15/2021 11 12 12/15/2022 10 12 12/15/2022 15 12 12/15/2020 18 11 12/15/2020 17 11 12/15/2022 23 11 12/15/2020 11 10 12/15/2020 14 10 12/15/2021 21 10 12/15/2021 19 10 12/15/2020 19 8 12/15/2020 8 8 12/15/2021 20 8 12/15/2022 12 8 12/15/2022 11 8 12/15/2020 6 7 12/15/2020 21 7 12/15/2020 22 7 12/15/2020 20 7 12/15/2020 15 7 12/15/2022 9 7 12/15/2022 7 7 12/15/2021 9 7 12/15/2022 21 7 12/15/2020 16 6 12/15/2020 0 6 12/15/2022 3 5 12/15/2022 20 5 12/15/2021 7 5 12/15/2021 0 4 12/15/2021 23 4 12/15/2022 2 4 12/15/2021 1 4 12/15/2022 22 4 12/15/2021 3 4 12/15/2021 8 4 12/15/2022 13 4 12/15/2020 4 3 12/15/2020 1 3 12/15/2022 5 3 12/15/2022 16 3 12/15/2022 1 3 12/15/2021 5 3 12/15/2020 12 2 12/15/2020 2 2 12/15/2020 3 2 12/15/2020 23 2 12/15/2021 6 2 12/15/2022 8 2 12/15/2022 0 2 12/15/2022 4 2 12/15/2021 4 2 12/15/2020 5 1 12/15/2020 7 1 12/15/2022 19 1

Databricks visualization. Run in Databricks to view.

In [0]:
from pyspark.sql.functions import desc
top_3_districts = sea_crime_df.groupBy("mcpp") \
     .count() \
     .orderBy(desc("count")) \
     .limit(3)
display(top_3_districts)

mcpp count DOWNTOWN COMMERCIAL 91534 CAPITOL HILL 72277 NORTHGATE 65864

Databricks visualization. Run in Databricks to view.

In [0]:
crime_eventOf3dist = sea_crime_df.filter(sea_crime_df.mcpp.isin('DOWNTOWN COMMERCIAL', 'CAPITOL HILL', 'NORTHGATE')).groupby('crime_against_category', 'time').count().orderBy('crime_against_category', 'time')
display(crime_eventOf3dist)

crime_against_category time count NOT_A_CRIME 2 1 NOT_A_CRIME 10 1 NOT_A_CRIME 11 1 NOT_A_CRIME 12 1 NOT_A_CRIME 19 1 PERSON 0 1828 PERSON 1 1869 PERSON 2 1913 PERSON 3 1177 PERSON 4 770 PERSON 5 629 PERSON 6 620 PERSON 7 746 PERSON 8 956 PERSON 9 1127 PERSON 10 1355 PERSON 11 1360 PERSON 12 1667 PERSON 13 1891 PERSON 14 1789 PERSON 15 1942 PERSON 16 1997 PERSON 17 2080 PERSON 18 1998 PERSON 19 1872 PERSON 20 1824 PERSON 21 1997 PERSON 22 1903 PERSON 23 1867 PROPERTY 0 3220 PROPERTY 1 2963 PROPERTY 2 2601 PROPERTY 3 2009 PROPERTY 4 1715 PROPERTY 5 1948 PROPERTY 6 4253 PROPERTY 7 5793 PROPERTY 8 7042 PROPERTY 9 8507 PROPERTY 10 9581 PROPERTY 11 9383 PROPERTY 12 10939 PROPERTY 13 11591 PROPERTY 14 10318 PROPERTY 15 10517 PROPERTY 16 9734 PROPERTY 17 9677 PROPERTY 18 8580 PROPERTY 19 7285 PROPERTY 20 7169 PROPERTY 21 6726 PROPERTY 22 5225 PROPERTY 23 4025 SOCIETY 0 1493 SOCIETY 1 1476 SOCIETY 2 1399 SOCIETY 3 890 SOCIETY 4 696 SOCIETY 5 546 SOCIETY 6 591 SOCIETY 7 697 SOCIETY 8 746 SOCIETY 9 858 SOCIETY 10 997 SOCIETY 11 1184 SOCIETY 12 1500 SOCIETY 13 1657 SOCIETY 14 1679 SOCIETY 15 1775 SOCIETY 16 1713 SOCIETY 17 1532 SOCIETY 18 1384 SOCIETY 19 1409 SOCIETY 20 1724 SOCIETY 21 2123 SOCIETY 22 2029 SOCIETY 23 1594

Databricks visualization. Run in Databricks to view.


#### 1. Crime Distribution by District
Top 3 High-Crime Districts:
Downtown Commercial: 91,534 crimes
Capitol Hill: 72,277 crimes
Northgate: 65,864 crimes
Advice:
Hotspot Policing: Focus on these high-crime districts with increased patrols, especially during peak crime hours.
Community Engagement: Work with local businesses and residents in these areas to develop community-based crime prevention strategies.
Infrastructure Improvements: Consider improving street lighting, installing surveillance cameras, and increasing police presence in these districts.
#### 2. Crime Distribution by Time of Day
Peak Crime Hours:
Property Crimes: Peak between 10 AM to 4 PM, with the highest number of crimes occurring around 1 PM.
Crimes Against Persons: Peak in the late afternoon and evening, with the highest number of crimes occurring between 5 PM to 9 PM.
Society-Related Crimes: Peak in the evening and late night, with the highest number of crimes occurring between 9 PM to 11 PM.
Advice:
Shift Adjustments: Adjust police shifts to ensure higher coverage during peak crime hours. For example, increase patrols in the late afternoon and evening when crimes against persons are more frequent.
Targeted Patrols: Focus on property crime prevention during the daytime, especially in commercial areas, and increase patrols in residential areas during the evening and night.
Public Awareness: Educate the public about safety measures during high-crime hours, such as avoiding walking alone at night and securing property during the day.